default.nix

```diff
@@ -1,5 +1,30 @@
-# This default.nix is designed to be passed directly nixpkgs with something like:
-#   import <nixpkgs> (import (builtins.fetchTarball https://github.com/input-output-hk/haskell.nix/archive/master.tar.gz))
-{ config   = import ./config.nix;
-  overlays = import ./overlays;
-}
\ No newline at end of file
+let
+  # Names nixpkgs pins (corrisponds to names of the nixpkgs/*.json files)
+  nixpkgs-names = ["release-18.08" "release-19.03" "release-19.09"];
+
+  # Name of the default nixpkgs-pin
+  default-nixpkgs-name = "release-19.09";
+
+  # These are from lib.attrset but we don't want to have to import a nixpkgs just to get them
+  nameValuePair = name: value: { inherit name value; };
+  genAttrs = names: f: builtins.listToAttrs (builtins.map (n: nameValuePair n (f n)) names);
+
+in {
+  # nixpkgsArgs is designed to be passed directly nixpkgs with something like:
+  #   import <nixpkgs> (import (builtins.fetchTarball https://github.com/input-output-hk/haskell.nix/archive/master.tar.gz)).nixpkgsArgs
+  nixpkgsArgs = {
+    config   = import ./config.nix;
+    overlays = import ./overlays;
+  };
+
+  inherit default-nixpkgs-name;
+  
+  # nixpkgs allows you to import haskell.nix with the default pinned nixpkgs
+  #   (import (builtins.fetchTarball https://github.com/input-output-hk/haskell.nix/archive/master.tar.gz)).nixpkgs {}
```

Comment:
Wouldn't you also need to pass the args? i.e. something like:
```
let haskell-nix = ...
in import haskell-nix.nixpkgs haskell-nix.nixpkgs-args
```

Reply:
I set it up so that `haskell-nix/nixpkgs/default.nix` adds them in by default.  It gets them using `(import ../.).nixpkgsArgs` (rather than having `./haskell-nix/default.nix` pass them in), so that when we have a local `haskell-nix` git repo we can just do `pkgs = import ./haskell-nix/nixpkgs {}` and get something usable.

Reply:
As a user, I'd find that a bit surprising. I'd assume that `nixpkgs` was just `nixpkgs` and that I'd need to add the overlay in as usual. And I can see two actual usage problems:
- I have to do different things if I'm using the `haskell.nix` pin vs my own pin
- If I want to import the pinned nixpkgs with some of my own nixpkgs args and some of the haskell.nix args (e.g. I need to add more overlays, or config), then I think I can't easily do that.

I think it's better if we give people the pieces and let them do the composition.

Reply:
Any thoughts on a good name for the variable to use in the examples and tests?  I'm not so keen on `haskell-nix` as it could easily be confused with `pkgs.haskell-nix`.

Also your example hints you expected `haskell-nix.nixpkgs` to require importing.  Is it ok that it is currently just a function that returns `pkgs`?

Reply:
Not off the top of my head. I agree that `haskell-nix` is a bit confusing.

My opinion (might just be me): I'd like anything that advertises itself to be a pin to be a *source*. That way you always know the interface: you import it and do whatever you normally do with that source. Otherwise you have to know that it's nixpkgs, but wrapped in some function that you now have to understand.

So I'd probably do something that amounted to:

```
{
    nixpkgs-pins = {
        "19.09" = builtins.fetchTarball ...
        ...
    };
    defaultNixpkgs = nixpkgs-pins."19.09";
    ...
}
```

Reply:
(TBH, I'd automatically call my imported `haskell.nix` repository `haskell-nix`, and I'd probably just accept it being a bit confusing.)

Reply:
I've gone with `haskellNix`, since some of docs have stuff like:

```
with import haskellNix.defaultNixpkgs haskellNix.nixpkgsArgs; haskell-nix.nix-tools
```